minidump/src/minidump.rs

```diff
@@ -116,6 +111,101 @@ pub trait MinidumpStream<'a>: Sized {
     fn read(bytes: &'a [u8], all: &'a [u8], endian: scroll::Endian) -> Result<Self, Error>;
 }
 
+/// Provides a unified interface for analyzing a [`MinidumpLinuxMaps`] and
+/// [`MinidumpMemoryInfoList`].
+///
+/// Currently an enum because there is no situation where you can have both,
+/// but this may change if the format evolves.
+#[derive(Debug, Clone)]
+pub enum UnifiedMemoryInfoList<'a> {
+    Maps(MinidumpLinuxMaps<'a>),
+    Info(MinidumpMemoryInfoList<'a>),
+}
+
+#[derive(Debug, Copy, Clone)]
+/// A [`UnifiedMemoryInfoList`] entry, providing metatadata on a region of
+/// memory in the crashed process.
+pub enum UnifiedMemoryInfo<'a> {
+    Map(&'a MinidumpLinuxMapInfo<'a>),
+    Info(&'a MinidumpMemoryInfo<'a>),
+}
+
+/// The contents of `/proc/self/mem` for the crashing process.
+///
+/// This is roughly equivalent in functionality to MinidumpMemoryInfoList.
+/// Use [`UnifiedMemoryInfoList`] to handle the two uniformly.
+#[derive(Debug, Clone)]
+pub struct MinidumpLinuxMaps<'a> {
+    /// The memory regions, in the order they were stored in the minidump.
+    regions: Vec<MinidumpLinuxMapInfo<'a>>,
+    /// Map from address range to index in regions. Use `MinidumpLinuxMaps::memory_info_at_address`.
+    regions_by_addr: RangeMap<u64, usize>,
+}
+
+/// A memory mapping entry for the process we are analyzing.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MinidumpLinuxMapInfo<'a> {
+    /// The first address this metadata applies to
+    pub base_address: u64,
+    /// The last address this metadata applies to
+    pub final_address: u64,
+
+    /// The kind of mapping
+    pub kind: MinidumpLinuxMapKind,
+
+    // FIXME: These could be bitflags but I'm not worried about it right now
+    pub is_read: bool,
+    pub is_write: bool,
+    pub is_exec: bool,
+    pub is_shared: bool,
+    pub is_private: bool,
+
+    // Fields in the format we ignore (not yet useful)
+    // * offset
+    // * dev
+    // * inode
+    _phantom: PhantomData<&'a u8>,
+}
+
+/// A broad classification of the mapped memory described by a [`MinidumpLinuxMapInfo`].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum MinidumpLinuxMapKind {
+    /// This is the main thread's stack.
+    MainThreadStack,
+    /// This is the stack of the thread with the given `tid`.
+    Stack(u64),
+    /// This is the process's heap.
+    Heap,
+    /// This is the "Virtual Dynamically-linked Shared Object".
+    Vdso,
+    /// This is an anonymous mmap.
+    AnonymousMap,
+    /// Some other special kind that we don't know/care about.
+    UnknownSpecial(String),
+    /// This is a mapped file/device at the given path.
+    File(String),
```

Comment:
This (and other places where you have `String`) might want to actually be `Vec<u8>` for correctness' sake since they're effectively `OsString` client-side, right? We tend to try to avoid dropping data from minidumps unless it's completely busted, so it would be good to be cautious here.

Is there a useful abstraction in Rust for "some other system's `OsString`? if not, I'd support a very small abstraction like:

```rust
struct DumpOsStr<'a>(&'a [u8]);

impl<'a> Into<Cow<'a, str>> for DumpOsStr<'a> {
    fn into(self) -> Cow<'a, str> {
      String::from_utf8_lossy(self.0)
    }
}
```

Reply:
mm this is a good point, I will ponder on this

Reply:
Yeah, on Linux that's definitely a `Vec<u8>` because traditional Unix paths have no encoding, they're just raw sequences of bytes. Nobody is crazy enough to use non-printable paths on their machines but we have to keep in mind that we're parsing untrusted data so we can't assume those are always valid strings.

Reply:
It's at least a tiny bit better than most other things since this data comes directly from the kernel and Breakpad includes it wholesale into the minidump, but the filenames do come from userland so I'm sure it's trivially easy for it to wind up as WTF-8 in practice if some code decides to write out a temp file with part of the name interpolated from other data and then mmap's it.

I guess you could fairly easily dump the maps streams from a random sample of Firefox Linux crash dumps and determine what percentage, if any, fail to parse as valid UTF-8.